dandi/core/digests/dandietag.py

```diff
@@ -0,0 +1,169 @@
+
+        if math.ceil(file_size / part_size) >= cls.MAX_PARTS:
+            part_size = math.ceil(file_size / cls.MAX_PARTS)
+
+        if part_size < cls.MIN_PART_SIZE:
+            part_size = cls.MIN_PART_SIZE
+
+        if part_size > cls.MAX_PART_SIZE:
+            part_size = cls.MAX_PART_SIZE
+
+        part_qty, final_part_size = divmod(file_size, part_size)
+        if final_part_size == 0:
+            final_part_size = part_size
+        else:
+            part_qty += 1
+        if part_qty == 1:
+            part_size = final_part_size
+        return cls(part_qty, part_size, final_part_size)
+
+    def __len__(self) -> int:
+        return self.part_qty
+
+    def __getitem__(self, index: int) -> Part:
+        if 1 <= index < self.part_qty:
+            return Part(
+                index, self.initial_part_size * (index - 1), self.initial_part_size
+            )
+        elif index == self.part_qty:
+            return Part(
+                index, self.initial_part_size * (index - 1), self.final_part_size
+            )
+        else:
+            raise IndexError(index)
+
+    def __iter__(self) -> Iterator[Part]:
+        offset = 0
+        for number in range(1, self.part_qty):
+            yield Part(number, offset, self.initial_part_size)
+            offset += self.initial_part_size
+        yield Part(self.part_qty, offset, self.final_part_size)
+
+
+class DandiETag:
+    REGEX = r"[0-9a-f]{32}-\d{1,5}"
+    MAX_STR_LENGTH = 38
+
+    def __init__(self, file_size: int):
+        self._part_gen: PartGenerator = PartGenerator.for_file_size(file_size)
+        self._md5_digests: List[Optional[bytes]] = [None] * len(self._part_gen)
+        self._next_index: int = 0
+
+    @property
+    def part_qty(self) -> int:
+        return len(self._part_gen)
+
+    @property
+    def complete(self) -> bool:
+        return self._next_index == self.part_qty
+
+    def get_parts(self) -> Iterator[Part]:
+        return iter(self._part_gen)
+
+    def get_next_part(self) -> Optional[Part]:
+        if self._next_index < self.part_qty:
+            return self._part_gen[self._next_index + 1]
+        else:
+            return None
+
+    def as_str(self) -> str:
+        if not self.complete:
+            raise ValueError("Not all part hashes submitted")
+        blob = b""
+        for d in self._md5_digests:
+            assert d is not None
+            blob += d
+        parts_digest = md5(blob).hexdigest()
+        return f"{parts_digest}-{len(self._md5_digests)}"
+
+    @classmethod
+    def from_file(
+        cls, path: Union[str, bytes, "os.PathLike[str]", "os.PathLike[bytes]"]
+    ) -> "DandiETag":
+        etag = cls(file_size=os.path.getsize(path))
+        with open(path, "rb") as f:
+            for part in etag.get_parts():
+                etag.update(f.read(part.size))
+        return etag
+
+    def add_digest(self, p: Part, part_digest: bytes) -> None:
```

Comment:
I would prefer to not expose in public interface methods exposing internal assumption (i.e. how `part_digest` is computed) and rather only operate on `block`s.  Please make `add_digest` and `add_next_digest` protected, and either add  `update_part(self, block: bytes, p: Part)` or make `update(self, block: bytes, p: Part=None)` so it would switch between the two depending either `p` is provided (that would also allow to centralize computing of md5() on a part).